Fix double exit payments on snl detach

Doy-lee opened the pull request:
The general premise here is to ensure that blockchain detach, detaches the SNL and SQL DB to the exact same height. This ensures that the systems rewind back and resync from the same starting point. Then, if the act of syncing the SNL affects the SQL DB like inserting new rows, this is handled correctly as the SQL DB is also at the exact same height as the SNL as to when the row was produced and any logic in the SQL DB that relies on that information is handled correctly.

The best way to do this is to just embed the SQL DB into the SNL and lock-step them together because they are inherently linked to each other (if you add a block to the SNL and _not_ the SQL DB then the ledger becomes incorrect). This is a hard requirement in the sequence of API calls you need to do when a block is added for correctness. That'd be a very large and disruptive change that I've not done. More reasonably I've deleted some public APIs on the SQL DB and merged them into the SNL to ensure the SQL is updated atomically with the SNL.

This leads to the following changes in the PR:

- SQL pop_blocks is deleted because the SNL does not support pop_blocks. However both SNL and SQL support blockchain detach by storing state at particular checkpoints so that is the method they use to handle pop blocks.
- The SQL and SNL backup exactly the same number of blocks for handling pop_blocks and reorgs. They use the shared variables from `network_config.h`
  - `HISTORY_ARCHIVE_INTERVAL`: The number of blocks between that we store a backup of the state, set to 10k blocks. This is the same value as we had before. These are stored for a rolling window of `HISTORY_ARCHIVE_KEEP_WINDOW` blocks. Archived state is for big re-orgs (typically invoked by using pop blocks manually).
  - `HISTORY_ARCHIVE_KEEP_WINDOW`: The rolling window of how many archives we keep. This is set to 2 years. Before this PR we kept them in perpetuity. Realistically this is not necessary and I've added this to bound the size of this over time.
  - `HISTORY_RECENT_KEEP_WINDOW`: How many blocks worth of state from the tip should be kept. This is for short re-orgs. This was previously a local variable in the SNL. This was lifted to be global so that it can be shared by the SQL DB.

Now both SNL and SQL DB use these variables to determine when they take a snapshot of their current state and in doing so both store the exact same number of backups that they can then detach to when requested and hence resync from the same starting point. The calculation and backup of state is done in the SNL and SQL individually which is ripe for bugs as the SNL has to backup using the binary serialisation system and the SQL DB using SQL queries.  To guard against bugs and in this and futurebugs I've added a bunch of sanity checks and asserts in various places to try and catch and recover if they go out of sync.

- When we commit rows from the SNL into the SQL DB (processing exit stakes) the row has a unique clause on the `(block_height, tx_index, contributor_index)` which outright prevents the SQL from inserting payment for a node twice if it somehow was made to process a block twice. This blocks the issue we're encountering now on stagenet where a rescan was re-processing exits into the SQL DB.
- Asserts to check that on rescan and block add, ONS, SQL and SNL are at the same height.
- On SNL detach, before detaching to the desired height it checks that the SNL and SQL DB both have backups of the height to detach to. If it doesn't, it'll walk back to find the height that they can rewind to before proceeding. If they can't agree then they both resync from scratch. This is also run on startup (we call `exec_detach_hooks` on `Blockchain::init`) which means that if at any point the daemon crashes, systems go out of sync, a restart should detect this and get them to resync so it's fairly more resilient over time to bugs.

Part of this change required was figuring out how the old code worked, what it did, e.t.c and so bug fixes and improvements were made in the pre-existing process that I believe make it more robust and understandable.

-  Fixed the backups of recent SNL state only storing quorums. We were calculating

```cpp
uint64_t const max_short_term_height =
            short_term_state_cull_height(hf_version, (m_state.height - 1)) + VOTE_LIFETIME +
            VOTE_OR_TX_VERIFY_HEIGHT_BUFFER;
```

Which was offsetting the height, and then again in the function offsetting the height argument to something OOB which would trigger the serialiser to only store quorums. That's fixed now having reviewed the serialisation routines. Removed most of the branches and the scary height calculations and made it just read top-to-bottom with explanations of each step.

- Removed various conversions back and forth from uint64_t -> int64_t in the oxen rescanning loop.
- Removed a bunch of SQL and SNL height checks at the start of the oxen rescanning loop. SNL and SQL accept heights below the hardfork that they're activated which means you can check that the heights match between the two and avoid having logic to clamp heights to the lowest common denominator if they are below the hardforks e.t.c
- Add loading metrics to the rescan which reveals that we're rescanning at an abysmal speed of less than 1 MiB/s. Room for improvement down the line now that the metrics are available.
- Remove STATE_CHANGE_TX_LIFETIME_IN_BLOCKS with VOTE_LIFETIME, they are the same value but also mean the same thing.
- Fix the `filter_and_sort` sorting by the contents of `pubkey_and_info` instead of just `pubkey`

Doy-lee committed:
Break up load blocks into subsystem into sections

- Store hardfork heights at the top of the function to reduce line
length and minimise the amount of line-breaking enforced by the
formatter. This breaks up the text-heavy noise in the height enumeration step.

- Separate the step where we calculate the minimum height of all the
subsystems which is the height to start loading blocks from to after
enumerating all the heights. This makes it very clear what heights we
need to run the min() function on and find the earliest height.

- Remove the vector accumulating heights. It's clearer now what heights to min()
now that the enumerating height and processing the height is separated.

- Avoid signed int math. Unsigned works fine by relying on well-defined
wrapping behaviour which removes the noisy static_casts in the main loop.

- Remove early exit checking on there being no blocks to check. The
bottom of the function had the exact same check. The loop is well
behaved for (total_blocks == 0) so we can just let the function proceed
rather than add another branch.

Doy-lee committed:
Fix SN sort_and_filter being able to sort by shared_ptr contents

In practice this never became a problem because our list is a unique set
of Ed25519 keys but for correctness we correctly compare _just_ the
Ed25519 contents for sorting.

Doy-lee committed:
Move repeated parameters in process_confirmed_event into struct

Doy-lee committed:
Store {block,tx,contrib} index in delayed_payments

We do a dupe insertion check against the {block,tx,contrib} index which
is guaranteed to be unique. Note we do _not_ check/store block hash
because our threat model here is that the block hash is valid by the
time this part of the code is met (e.g. it passed all of blockchain and
SNL validation essentially).

To do this we have to pipe through some new information from the SNL
when it calls into SQL DB to insert a delayed payment (i.e. return of
stake). This makes it so that essentially there is 1 row for each
returned stake as they are uniquely identified by the {block,tx,contrib}
index.

Doy-lee committed:
Correctly share STORE_LONG_TERM_STATE_INTERVAL in SNL and SQL

Doy-lee committed:
Rename STORE_LONG_TERM_STATE_INTERVAL/STORE_RECENT_REWARDS -> HISTORY_ARCHIVE_INTERVAL/HISTORY_KEEP_RECENT_WINDOW

Doy-lee committed:
Use HISTORY_KEEP_RECENT_WINDOW in SNL&SQL to unify stored backups

Incremental steps to getting the the SQL and SNL to handle
blockchain_detached in the same manner. They both need to detach to
exactly the same height so they must also store backups to the same
degree on both sides so that one subsystem doesn't desync on detach.

If they desync, we risk the 2 systems producing inconsistent rewards
which breaks BLS aggregation.

Doy-lee committed:
Rewrite state serialization to execute linearly

- Short term serialisation was broken because the height that was
calculated as the cutoff height to serialise was incorrectly adding a
offset to the current height before being subtracted again such that all
the SNL states that were serialised only serialised their quorums.

This means that every time we reorg, 1 or 2 blocks, this triggers a full
rescan from the last 10k-th block interval. This was hidden in the code
because the logic was hard to read and modify.

- The serialisation routine for some reason was also trying to serialise
the quorums at a cut-off height, but, when you process a block into the
SNL you _already_ cull history that is too old. Hence when you go to
serialise, the list of states has already been pruned, there's no need
to have logic to prune it again. All we have to do is serialise it to
disk. That simplifies the serialisation routine which unveiled the bug
in the above bulletpoint.

- Remove tricky logic that migrated the short-term window into the
archive window. This had a bunch of checks embedded in a loop with some
arithmetic to calculate some offsets from the height. This is not
necessary if we just serialise the state on block add by determining if
the current state falls under the criteria to be serialised into the
archive or not. This simplifies the loop, de-nests the algorithm and
hence reduces the cognitive load required to grok what's happening.

- Fix an issue where SNL::load() calls reset(false). Reset was clearing
out the SQL DB causing a full-rescan on startup every time even if the
SNL and SQL DB's were in sync.

Doy-lee committed:
Add throughput stats to subsystem rescan

Doy-lee committed:
Initiate one SQL transaction at start of upgrade_schema

Doy-lee committed:
Setup SQL and SNL to store the same amount of archive backups

Unifies the amount of archive/recent state kept by the SQL and SNL DB.
This means that when the systems detach, they are able to detach to the
same height as each other and then able to resync the chain from the
same starting point.

Syncing from the same starting point is paramount to ensuring that the
same results are derived as loading blocks into the SNL has side effects
on the SQL DB.

Doy-lee committed:
Ensure SNL and SQL cull/keep the same heights

Doy-lee committed:
Serialize SNL state before block is processed

This ensures that the first SNL state we store in our recent history
matches the first SQL state that the SQL DB stores. This means a slight
adjustment in the SNL serialisation code in that it has to serialise
itself, 'm_state' in tandem with 'm_transient.state_history'.

Doy-lee committed:
Add wrapper around blockchain detach to detect errors

The wrapper allows us to check and assert if the detach popped us to the
right block for testing and verifying that the detach is working.

Small rewrite of the detach logic in SNL to read top-to-bottom with
minimal branching or early returns.

Doy-lee committed:
Remove pop_blocks API from SQLite and correctyl restore from archive/recent tableL

Doy-lee committed:
Make SNL detach call directly into SQL detach to lockstep the detach step

Doy-lee committed:
Allow detach of SNL and SQL to lower height than requested

We detach to a height that is supported by both SNL and SQL by checking
the backups available on both before proceeding. This ensures in
exceptional cases if the SQL side is missing a backup, the SNL doesn't
go ahead and rewind itself. This would cause a desync that can only be
remedied by noticing it, and then restarting the daemon.

Instead the 2 double check what is a safe height they both support and
do that patching up that particular avenue of breakage.

Doy-lee committed:
Format non-complying code

Doy-lee committed:
Cleanup pass over PR, fix comments/grammar, remove fakechain SQL check

Seems to run fine without the FAKECHAIN check. I checked the unit tests
and they use a real SQL DB instance so there should be no issue _not_
gating the code from executing the code.

Doy-lee committed:
Serialize SNL state before updating

Otherwise we store the latest height twice, instead of once

Doy-lee committed:
Detach ONS in the detach hook after all blocks popped

This matches behaviour with the other detach calls that detach after all
blocks have been popped. This gets rid of the detach function in ONS
which just called prune_db which we do so directly in the detach hook.

Doy-lee committed:
Ensure that SQL block is added instead of early exiting in SNL block add

Doy-lee committed:
Remove outdated comment about SQL DB not being present in FAKECHAIN

Doy-lee committed:
ONS check for nullptr, unit tests invokes w/o initialising DB

Doy-lee committed:
Handle unit-tests asserting due to missing txin_gen in miner_tx w/ FAKECHAIN on startup

Doy-lee committed:
Check if the SQL DB is present for unit tests

Doy-lee committed:
Add integration test for delayed_paments rollback

Doy-lee committed:
Verify batch db info height was undone as well in integration tests

Doy-lee committed:
Use special local-devnet deploy script to avoid exit-too-early by 2hrs problem

The devnet version of the script deploys a contract that inherits the
mainnet ServiceNodeRewards contract but instead overrides the minimum
exit time. The contract currently rejects exits that occur within the
first 2hrs of the node registering.

We don't have an easy way to get the nodes to fast-fwd their clocks by
2hrs to sign a signature a time in the future.

Doy-lee committed:
Remove subtract_sn_rewards from db_sqlite

We don't support popping a block from the SQL DB anymore, only detach to
a recent height which is supported by the SNL, ONS and SQL DB. This is
required for deterministic behaviour on rewind. Since that's the case we
can remove the subtract code from the DB and simplify.

Doy-lee committed:
Fix bool return in add_sn_rewards being unused in db_sqlite

This function always returns true meaning downstream checks on this call
always succeed. This function can fail at the SQL insertion step.
Predominantly, the pattern in the SQL DB is to use exceptions as the
error mechanism and the one function that does call `reward_handler` is
wrapped in a try-catch block.

I've opted to remove the boolean return and instead annotate that the
function throws. This simplifies the calling code in that it no longer
has to do a success check (which was redundant since true was always
returned) and instead optimistically executes and defers to the
try-catch to handle errors.

Doy-lee committed:
Have transition path for syncing from oxen-stable DB

tewinget commented on src/blockchain_db/sqlite/db_sqlite.cpp:
```diff
@@ -356,37 +426,55 @@ void BlockchainSQLite::update_height(uint64_t new_height) {
+    }
 
-      DELETE FROM batched_payments_accrued;
+    // NOTE: When we detach we are rewinding to an old state. The accrued table
```
I'm not sure this is true.  I think because of the trigger `batch_payments_delete_empty` that an archive at an older height can have more rows, as someone fully unstaking and cashing out will have their addresses' rows removed.

Doy-lee replied:
Oh that's interesting. I didn't notice that trigger. Though if someone fully unstakes and cashes out, that is an additive operation to their row. Their exit stake gets added into their row's `amount` counter. 

But, now that I've removed the popping of blocks from the DB which would subtract rewards on a block-by-block basis (because popping breaks when you remove a delayed payment), there's no operation you can do on the DB to subtract. 

On detach you delete all the rows and restore all the old rows, and to be present in a DB snapshot you must have a non-zero rewards amount so you can't ever encounter a row with 0 rewards for this trigger to execute?

If that's the case, then I'll push a commit to remove the trigger if you don't see an issue with that.

tewinget replied:
by "cashing out" I meant claiming the entirety of their rewards, forgetting that we indeed no longer subtract from that table after HF21.  Up until that point, I suppose when an address' batch payment goes through its row would zero out, so there's maybe a window of time where we have to care.

Doy-lee replied:
I have removed this check

tewinget commented on src/blockchain_db/sqlite/db_sqlite.cpp:
```diff
@@ -283,41 +320,74 @@ void BlockchainSQLite::upgrade_schema() {
+        DROP TRIGGER IF EXISTS delayed_payments_prune;
+        CREATE TRIGGER delayed_payments_prune AFTER UPDATE ON batch_db_info
+        FOR EACH ROW BEGIN
+            DELETE FROM delayed_payments WHERE payout_height < NEW.height;
```
I *think* this needs to be `WHERE entry_height < NEW.height`.  As-is, a re-org right after a delayed payment is paid out would revert the payout but no longer have the delayed payment entry in this table.

Doy-lee replied:
So you've actually uncovered a bunch of bugs having reviewed this after your findings. We need 2 triggers here, one from blockchain detached and one for when a block is added.

This one you've pointed out was originally written to handle the `on_block_add` case. When we add a block, we process all the transactions in the block, give users back their stake (delayed payment) if it's queued for this block, then we execute this trigger which deletes these delayed payment that has been executed.

Then there's the situation where the blockchain reorgs/detaches. In that case we need the condition you've mentioned but I believe with `>` instead of `<` which is that `entry.height > NEW.height`. Then when oxen replays the chain it'll recreate the delayed payments when it passes at the `entry.height`.

So this stuff's pretty confusing, let me know if what I've summarised is correct. Also note that the SQL stores the chain height not the chain length, i.e. it stores the block number that it last executed so that's the reasoning for `<=` and `>` and so forth in the commit.

tewinget replied:
> When we add a block, we process all the transactions in the block, give users back their stake (delayed payment) if it's queued for this block, then we execute this trigger which deletes these delayed payment that has been executed.

Since we're not snapshotting this table, the chain of events `add block -> payout -> delete row because payout made -> "pop" block (well, some small amount of detach)` will lead to that delayed payment not happening if the detach lands between `entry.height` and `payout.height`, no?  Which...hmm, we might actually need to snapshot it for that reason now that I think about it.  Like, if we want to be able to say "those DBs got out of sync, rewind them to synchronize" without going from the beginning of time, I'm pretty sure we can rewind to such a situation.

Doy-lee replied:
Oh you're right. I wonder if this is the thing breaking the blockchain syncing when we sync dev onto stable because right now it does a big rescan from some earlier checkpoint, the old trigger was this

```
        CREATE TRIGGER delayed_payments_prune AFTER UPDATE ON batch_db_info
        FOR EACH ROW BEGIN
            DELETE FROM delayed_payments WHERE payout_height < (NEW.height - 10000);
        END;
```

I was like what's up with the magic number, that'd break in the way you describe if you go further back than 10k blocks. Well .. I didn't want to but I guess I'll have to do the same archive table thingy with the delayed payments.

Doy-lee replied:
So I've gone and done this and I had to shuffle some more code around but I think it's in a better place now. I merged a bunch of triggers together, so there's one trigger for block-reorg, one trigger to save rows on block add and so forth. So it's less scattered all over the place.

This broke a couple of things so I had to go and debug them and issue some fixes where more bugs were found. I made the integration test more complicated to check for this `add block -> payout -> delete row because payout made -> "pop" block (well, some small amount of detach)` situation, took a bit of time due to how long it takes to run those tests, iterate and verify.

tewinget commented on src/cryptonote_core/blockchain.cpp:
```diff
@@ -460,37 +459,55 @@ bool Blockchain::load_missing_blocks_into_oxen_subsystems(const std::atomic<bool
+    auto end = clock::now();
+
+    if (total_blocks > 0) {
+        // NOTE: Check that all subsystems ended up synchronised to the same height
```
I wonder if these checks should either throw or reset these databases and re-call this function?  Either way, probably good to also check the heights are in sync at the beginning of this function as well (or 0 if the chain isn't to their relevant height yet I guess).  If that's checked already outside where this is called and I haven't gotten there yet, disregard.

Doy-lee replied:
Yeah this function is somewhat resilient to mismatched heights when it's called. The SNL is responsible for adding the block to the SQL DB internally. When it does that it checks that they are in sync. That throws if it isn't.

The only ways I could think of for this to go out of sync was

```
    // These 2 systems can go out of sync, though generally only in exceptional
    // cases, for example:
    //
    //  If a block was successfully added to the SNL but an exception was
    //  thrown when trying to add to the SQL DB (maybe a logic bug, DB is
    //  corrupted or DB is not in a writable state).
    //
    //  The daemon is terminated in the middle of updating the SNL or SQL DB.
    //
    //  The SQL DB was deleted or SNL was reset.
```

What happens then is the SNL will throw, it'll abort whatever operation just happened. On startup, if we detect that the heights are not in synced, it'll detach to the lowest common shared height and then call this function to resync them. So in practice we should be somewhat resilient to out-of-sync DBs.

tewinget replied:
ok cool.  That logic is scattered around enough that I thought it prudent to ask just in case.

tewinget commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -77,23 +77,21 @@ namespace feature = cryptonote::feature;
+    // 360 is derived via (6 * VOTE_LIFETIME) where VOTE_LIFETIME is 60 blocks,
+    // e.g. Keep atleast the last 360 blocks worth of votes (which is short for
+    // state change TXs in this codebase)
+    assert(KEEP_WINDOW >= 360 && "Not enough recent backups for blink quorum retrieval!");
```
Comment still says "compilation failure" from when this was a `static_assert`

Doy-lee replied:
Fixed

tewinget commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -1735,26 +1725,31 @@ bool service_node_list::state_t::process_confirmed_event(
             proof.store(key, sn_list->blockchain);
         }
-        if (my_keys && my_keys->pub == key)
+        if (confirm.my_keys && confirm.my_keys->pub == key)
```
It feels a little weird to have bundled `my_keys` into this struct, but it's not a big deal.

tewinget commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -3286,10 +3255,41 @@ void service_node_list::state_t::update_from_block(
+                            }
+                        }
+                        if (!found)
+                            throw;
```
naked throw outside of a catch block

Doy-lee replied:
Doesn't it get caught by the one just a couple of lines below it? Or is my cpp-foo too weak and that's not how that works.

tewinget replied:
Wasn't sure about this myself, but a cursory search found [this comment](https://stackoverflow.com/a/2288727) which quotes the standard:

> If no exception is presently being handled, executing a throw-expression with no operand calls `terminate()`.

If you were to `throw some_exception{"bla"};` here, it would indeed be caught just below.

Doy-lee replied:
TIL, I have pushed a change to not do that.

tewinget commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -3402,65 +3402,117 @@ void service_node_list::state_t::update_from_block(
-        m_transient.state_history.erase(m_transient.state_history.begin(), end_it);
+    }
+
+    // NOTE: Store quorums form this height if requested (m_store_quorum_history is a CLI flag)
```
form -> from

Doy-lee replied:
Fixed

Doy-lee committed:
Fix broken 'make_archive'

Make archive was calculating the wrong height to prune a DB by when
a block is processed and it's a height eligible for archiving rows. It
calculates

```
height < (NEW.HEIGHT - HISTORY_ARCHIVE_KEEP_WINDOW) % HISTORY_ARCHIVE_INTERVAL
```

The RHS calculation is always going to produce a height in the range
[0, HISTORY_ARCHIVE_INTERVAL) which is not going to give us a rolling window of
the past HISTORY_ARCHIVE_KEEP_WINDOW archives. I've updated this to

```
height < (NEW.HEIGHT - HISTORY_ARCHIVE_KEEP_WINDOW)
```

There might be some archives that slip through if the KEEP_WINDOW and
ARCHIVE_INTERVAL are not divisible by each other but this should not be
a big deal.

Doy-lee committed:
Ensure we have 2 triggers for block add and detach for delayed payments

Doy-lee committed:
Fix outdated assertion comment, form->from typo

Doy-lee committed:
Fix linting error in SNL w/ data_for_serialization version

Doy-lee committed:
Avoid naked throw in TX lookup after confirmation

Doy-lee committed:
Fix broken claimRewards in test and add more syncing diagnostics

Doy-lee committed:
Store delayed payment history, fix delayed payments being applied too early

Delayed payments with a delay blocks of 1 was being applied to the SQL
DB before its SQL height was updated. The delayed payment was being committed at
the current height of the SQL DB (which was behind by one block as it
had not yet processed the block yet).

The block would be added to the DB updating the height of the SQL DB and
cause the triggers to execute and remove the delayed payments scheduled
for the current height of the SQL DB (e.g. the new one we just added)
before it can be paid out to the user in question.

Doy-lee committed:
Fix localdev test detection of delayed payments

Doy-lee committed:
Merge the delete trigger on detach into the re-org trigger

Doy-lee committed:
Fix incorrect stake check and bad estimation of stake return

Doy-lee committed:
Adhere to linter in SNL.cpp

Doy-lee committed:
Remove row restored assert, there's a brief period in <=HF19 where this is not the case

Doy-lee committed:
Fir linting error in db_sqlite.cpp, extraneous newline

Doy-lee committed:
Fix repeated insertion of delayed payments

Doy-lee committed:
Fix clear_recent_and_archive trigger not being reset every run
